core/message_queue.go

```diff
@@ -0,0 +1,131 @@
+package core
+
+import (
+	"sync"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/pkg/errors"
+)
+
+// MessageQueue stores an ordered list of messages (per actor) and enforces that their nonces form a contiguous sequence.
+// Each message is associated with a "stamp" (an opaque integer), and the queue supports expiring any list
+// of messages where the first message has a stamp below some threshold. The relative order of stamps in a queue is
+// not enforced.
+// A message queue is intended to record outbound messages that have been transmitted but not yet appeared in a block,
+// where the stamp could be block height.
+// MessageQueue is safe for concurrent access.
+type MessageQueue struct {
+	lk sync.RWMutex
+	// Message queues keyed by sending actor address, in nonce order
+	queues map[address.Address][]*queuedMessage
+}
+
+type queuedMessage struct {
+	msg   *types.SignedMessage
+	stamp uint64
```

Comment:
Why not just call this `blockHeight`? That's the intent, and calling it `stamp` when it's actually a block height will be confusing.

Reply:
I hear you and it's worth talking about it. I intend this data structure to be self contained, in that it implemented the interface and semantics described in its comments. While it does have an intended use case, it's correctness need not depend on that. 

It wasn't an obvious answer to use block height for the timeout. If we change our mind, nothing in this structure should need to change.